### PR TITLE
Improve test coverage and refactor

### PR DESCRIPTION
**Pull request by fbiville**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+gen-mocks: ## generate mocks
+	go generate ./...
+
+build: gen-mocks $(OUTPUT) ## build the project binary
+
+test: ## run the project tests
+	go test -v ./...
+
+install: build ## copy the binary to GOBIN
```
Comment (ericbottard) [inline Makefile]: Is this really necessary?

File: cmd/provisioner/main.go [hunk 2]
```diff
@@ -17,103 +17,46 @@
...
-
 	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-		if r.Method == http.MethodPut {
-			handlePut(w, r)
-		} else {
+		if r.Method != http.MethodPut {
 			w.WriteHeader(http.StatusMethodNotAllowed)
+			return
 		}
+		handleCreationRequest(broker, gateway, w, r)
```
Comment (ericbottard) [inline cmd/provisioner/main.go:43]: Not always a "Creation"...

File: cmd/provisioner/main.go [hunk 3]
```diff
@@ -17,103 +17,46 @@
...
-		_, _ = fmt.Fprintf(os.Stderr, "Reported successful topic %q\n", topicName)
-	}
-}
-
-type result struct {
-	Gateway string `json:"gateway"`
-	Topic   string `json:"topic"`
+	}()
+	requestHandler := &handler.TopicCreationRequestHandler{KafkaClient: kafkaClient, Gateway: gateway, Writer: os.Stderr}
+	requestHandler.GetHandlerFunc()(writer, request)
```
Comment (ericbottard) [inline cmd/provisioner/main.go:63]: Why introduce a function that returns a function here?

File: pkg/provisioner/handler/handler_test.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+	It("returns 200 if the topic already exists", func() {
+		fakeKafkaClient.TopicExistsReturns(true, nil)
+
+		creationHandlerFunc.ServeHTTP(responseRecorder, request)
+
+		Expect(responseRecorder.Code).To(Equal(http.StatusOK),
+			fmt.Sprintf("Expected %d after topic creation request but got %d", http.StatusOK, responseRecorder.Code))
+		Expect(responseRecorder.Body.String()).To(MatchJSON(
+			fmt.Sprintf("{"+
```
Comment (ericbottard) [inline pkg/provisioner/handler/handler_test.go:51]: Use backticks to not have to escape double quotes

File: pkg/provisioner/kafka/client.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	CreateTopic(topicName string) error
+	Close() error
+}
+
+type kafkaClient struct {
+	Admin   sarama.ClusterAdmin
+}
+
+func NewKafkaClient(brokerAddress string, logWriter io.Writer) (KafkaClient, error) {
+	sarama.Logger = log.New(logWriter, "[Sarama] ", log.LstdFlags)
```
Comment (ericbottard) [inline pkg/provisioner/kafka/client.go]: This is reassigning a global var multiple times and possibly concurrently.
Better left globally in main()

File: tools/tools.go [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// +build tools
+
+package tools
+
+import (
+	_ "github.com/maxbrunsfeld/counterfeiter/v6"
+)
+
+// This file imports packages that are used when running go generate, or used
```
Comment (ericbottard) [inline tools/tools.go]: Can't we `go get` it in the makefile, like we do for other tools in eg system?

File: cmd/provisioner/main.go (hunk 3 above)
Comment (fbiville) [reply cmd/provisioner/main.go:63]: That way, it adheres to `HandlerFunc` and can be directly tested with `myHandler.ServeHTTP`

File: pkg/provisioner/kafka/client.go (hunk 5 above)
Comment (fbiville) [reply pkg/provisioner/kafka/client.go]: It's reassigning to the same value, though, isn't it?

File: cmd/provisioner/main.go (hunk 2 above)
Comment (fbiville) [reply cmd/provisioner/main.go:43]: from the requester point of view, it is, though, isn't it?

File: pkg/provisioner/kafka/client.go (hunk 5 above)
Comment (fbiville) [reply pkg/provisioner/kafka/client.go]: (note that the global vars have been removed)

Comment (ericbottard) [reply pkg/provisioner/kafka/client.go]: It's a global var defined by sarama itself.
Need to move the assignment back to main()

Comment (fbiville) [reply pkg/provisioner/kafka/client.go]: oh, sorry, I thought you were referring to the broker address!
gotcha, moving it back to main

File: cmd/provisioner/main.go (hunk 2 above)
Comment (ericbottard) [reply cmd/provisioner/main.go:43]: No, hence the use of PUT. Will rename

File: cmd/provisioner/main.go (hunk 3 above)
Comment (ericbottard) [reply cmd/provisioner/main.go:63]: I'll word it the other way then: why create a struct? It does not have any other method and just serves as a trampoline for creating the handler

Why not 
```
   NewHandler(KafkaClient, Gateway,Writer) HandlerFunc
```

I'll merge as-is for now, but I think we should revisit.

**Comment by ericbottard** [comment]
Merged as c824d0b

**Pull request closed without merging**